### ISSUE TITLE
vmagent: wrong relabeling via -remoteWrite.urlRelabelConfig

### ISSUE THREAD

**Issue by Allineer** [VictoriaMetrics/VictoriaMetrics#599]
**Describe the bug**

`-remoteWrite.urlRelabelConfig`:

```
- source_labels: [job, agent]
  regex: postgres;gate
  replacement: true
  target_label: __keep

- source_labels: [__keep]
  regex: true
  action: keep
```

I see metrics for other jobs and agents in the database:

<img width="1913" alt="Снимок экрана 2020-07-02 в 15 34 29" src="https://user-images.githubusercontent.com/1168049/86359359-94b25000-bc79-11ea-906c-6055008bff03.png">

<img width="1913" alt="Снимок экрана 2020-07-02 в 21 45 25" src="https://user-images.githubusercontent.com/1168049/86397900-5f275a00-bcad-11ea-86b9-81ef283fbf50.png">

another `-remoteWrite.urlRelabelConfig`:

```
- source_labels: [job, agent]
  regex: extra;shkoder
  replacement: true
  target_label: __keep

- source_labels: [__keep]
  regex: true
  action: keep
```

<img width="1913" alt="Снимок экрана 2020-07-02 в 15 30 25" src="https://user-images.githubusercontent.com/1168049/86358952-02aa4780-bc79-11ea-9bc3-ae7278c13599.png">

**Expected behavior**

Correct filtering.

**Version**

```
vmagent-20200701-124503-tags-v1.37.4-0-g8da3f773a and v1.37.3
```

**Comment by valyala** [comment]
@Allineer , could you add `{job!="victoria-metrics"}` filter to the query from the first screenshot in order to see metrics for other jobs? Make sure you didn't enable self-scraping on VictoriaMetrics side with `-selfScrapeInterval`, because the self-scraped metrics aren't filtered by `vmagent`.

**Comment by Allineer** [comment]
@valyala, no other metrics atm.

<img width="1913" alt="Снимок экрана 2020-07-03 в 07 55 21" src="https://user-images.githubusercontent.com/1168049/86433177-a5f36f00-bd02-11ea-8dbc-eb4cce6a0b59.png">

**Comment by Allineer** [comment]
Something new.
Third database. Postgres job is in testing stage, so it's metrics routed only for @testing database. But..

<img width="1913" alt="Снимок экрана 2020-07-03 в 14 04 24" src="https://user-images.githubusercontent.com/1168049/86463493-4adb6f80-bd36-11ea-8ce8-60d02622ab53.png">

**Comment by Allineer** [comment]
Updates:

<img width="1913" alt="Снимок экрана 2020-07-06 в 07 57 46" src="https://user-images.githubusercontent.com/1168049/86557336-85bfec00-bf5e-11ea-89ad-9375423c041a.png">
<img width="1913" alt="Снимок экрана 2020-07-06 в 07 58 17" src="https://user-images.githubusercontent.com/1168049/86557348-89537300-bf5e-11ea-9b4e-01d59221f8d3.png">

**Comment by Allineer** [comment]
magenta's logs:

```
2020-07-07T07:53:00.466Z	error	VictoriaMetrics/app/vmagent/remotewrite/client.go:237	couldn't send a block with size 19520 bytes to "http://127.0.0.1:8427/api/v1/write": write tcp4 127.0.0.1:39608->127.0.0.1:8427: write: broken pipe; re-sending the block in 2.000 seconds
2020-07-07T08:08:00.472Z	error	VictoriaMetrics/app/vmagent/remotewrite/client.go:237	couldn't send a block with size 19512 bytes to "http://127.0.0.1:8427/api/v1/write": write tcp4 127.0.0.1:43568->127.0.0.1:8427: write: broken pipe; re-sending the block in 2.000 seconds
2020-07-07T08:13:00.466Z	error	VictoriaMetrics/app/vmagent/remotewrite/client.go:237	couldn't send a block with size 19530 bytes to "http://127.0.0.1:8427/api/v1/write": write tcp4 127.0.0.1:44794->127.0.0.1:8427: write: broken pipe; re-sending the block in 2.000 seconds
```

I have 20-30 messages with this error daily.
:8427 is the `victoria-metrics@etc` instance.

**Comment by Allineer** [comment]
And WOW!

<img width="1913" alt="Снимок экрана 2020-07-07 в 11 35 34" src="https://user-images.githubusercontent.com/1168049/86749571-06a5e300-c046-11ea-875d-ae2a66c9a4e3.png">

**Comment by valyala** [comment]
The last graph looks really weird - `scrape_samples_post_metric_relabeling` metric shouldn't contain `type` label unless this label is explicitly added via relabeling (I assume this isn't your case according to relabeling configs posted in the initial message).

I'm still investigating this issue... It looks like some data race or improper cleanup of reused structures during relabeling.

**Comment by Allineer** [comment]
I have a really extensive "promscrape" config with a lot of relabeling.
And even if I assume that I add the “type” label as a result of erroneous re-labeling, then why does it appear only twice, for a short period and without my participation?

**Comment by valyala** [comment]
@Allineer , the commit a730b3f6a11b44e46cb8ea66a5f2f1db558997d3 should fix the issue. Could you try building `vmagent` from this commit according to [these instructions](https://victoriametrics.github.io/vmagent.html#how-to-build-from-sources) and verify whether the issue is gone?

**Comment by Allineer** [comment]
@valyala 

```
2020-07-12T14:07:11.396Z	info	lib/promscrape/scraper.go:303	static_configs: added targets: 24, removed targets: 0; total targets: 24
panic: interface conversion: interface {} is []prompbmarshal.TimeSeries, not *[]prompbmarshal.TimeSeries

goroutine 33 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.(*remoteWriteCtx).Push(0xc000023fc0, 0xc0004fc600, 0x5, 0x8)
	/tmp/VictoriaMetrics/app/vmagent/remotewrite/remotewrite.go:218 +0x46c
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.Push(0xc0004d4718)
	/tmp/VictoriaMetrics/app/vmagent/remotewrite/remotewrite.go:145 +0xfb
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*scrapeWork).scrapeInternal(0xc0004d4600, 0x1734359e903, 0x0, 0xf1e8c0)
	/tmp/VictoriaMetrics/lib/promscrape/scrapework.go:230 +0x3f0
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*scrapeWork).scrapeAndLogError(0xc0004d4600, 0x1734359e903)
	/tmp/VictoriaMetrics/lib/promscrape/scrapework.go:183 +0x4d
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*scrapeWork).run(0xc0004d4600, 0xc0002d0f60)
	/tmp/VictoriaMetrics/lib/promscrape/scrapework.go:157 +0x3b7
github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*scraperGroup).update.func1(0xc000092140, 0xc0004d4600, 0xc0004b1988)
	/tmp/VictoriaMetrics/lib/promscrape/scraper.go:284 +0x69
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape.(*scraperGroup).update
	/tmp/VictoriaMetrics/lib/promscrape/scraper.go:282 +0x3e2
2020-07-12T14:07:13.804Z	info	lib/logger/flag.go:11	build version: vmagent-20200712-133719-heads-master-0-ga730b3f6
```

**Comment by valyala** [comment]
@Allineer , thanks for testing! The panic should be fixed in the commit b442a42d8e798df6c74ee9934eeb2983f5128ab9

**Comment by Allineer** [comment]
Already testing...

**Comment by valyala** [comment]
@Allineer , FYI, all the bugfix commits mentioned in this issue have been included in [v1.38.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Comment by Allineer** [comment]
It's working! Thanks!
